### PR TITLE
fix: ensure entity_id is included in transaction events for reliable entity resolution

### PR DESCRIPTION
**Pull request by mabubakr1113**
This PR fixes the missing `entity_id` issue in indexer transaction events.

Previously, some transaction events were created without `entity_id`, which made it difficult to reliably resolve the related Trust Registry, Credential Schema, or Permission from the event data.

## What Changed

- Added `entity_id` extraction from `tx_response.events[].attributes[]`.
- Avoided using message content/payload as the source of truth for `entity_id`.
- Added proper event-to-ID mapping for:
  - Trust Registry events
  - Credential Schema events
  - Permission events
  - `SelfCreatePermission`
  - Permission VP lifecycle events
- Added fallback support for generic `id` where chain events emit non-standard ID keys.
- Prevented unrelated IDs such as `schema_id`, `trust_registry_id`, `tr_id`, and `validator_perm_id` from being used as Permission `entity_id`.
- Added conflict handling to avoid storing incorrect IDs when multiple conflicting candidates are found.
- Updated insert conflict merge to safely fill missing `entity_id` and `payload.entity_id`.
- Added unit tests for supported event mappings and edge cases.

## Why

Including `entity_id` directly in `indexer_events` allows consumers to reliably query and resolve the related entity without extra lookups or unsafe guessing.

## Testing

Added/updated unit tests covering:

- Create/Update/Archive Trust Registry events
- Create/Update/Archive Credential Schema events
- Create Root Permission
- Start Permission VP
- Self Create Permission
- Permission lifecycle events
- Generic `id` fallback
- Ignoring unrelated IDs
- Conflicting ID candidate handling

### CODE REVIEW

File: src/services/api/indexer_events_query.ts [hunk 1]
```diff
@@ -205,144 +180,354 @@ function collectDids(value: unknown, out: Set<string>): void {
...
-    .where("height", "<=", height)
-    .orderBy("height", "desc")
-    .first();
-  trustRegistrySnapshotCache.set(cacheKey, { expiresAt: now + DID_QUERY_CACHE_TTL_MS, value });
-
-  const row = await value;
-  addDid(dids, (row as any)?.did);
-  addDid(dids, (row as any)?.controller);
+  return undefined;
 }
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/services/api/indexer_events_query.ts]: Yep good spot. It’s unused and duplicates the conflict warning we already emit in resolveEntityIdFromTxLocalData, so I removed it.

File: src/services/api/indexer_events_query.ts [hunk 2]
```diff
@@ -205,144 +180,354 @@ function collectDids(value: unknown, out: Set<string>): void {
...
+      idKeys: ["permission_id", "perm_id", "id"],
+    },
+    set_permission_vp_to_terminated: {
+      eventTypes: ["set_permission_vp_to_terminated"],
+      idKeys: ["permission_id", "perm_id", "id"],
+    },
+    set_permission_vp_to_cancelled_last_request: {
+      eventTypes: ["cancel_permission_vp_last_request", "set_permission_vp_to_cancelled"],
+      idKeys: ["permission_id", "perm_id", "id"],
+    },
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/services/api/indexer_events_query.ts]: You’re right. The normalized action is cancel_permission_vp_last_request, so that mapping never matched. I renamed the key so the filter actually applies.

File: src/services/api/indexer_events_query.ts [hunk 3]
```diff
@@ -205,144 +180,354 @@ function collectDids(value: unknown, out: Set<string>): void {
...
+      VeranaPermissionMessageTypes.StartPermissionVP,
+    ].includes(messageType as any)
+  ) {
+    await debugTxAttributesForMissingEntityId(row);
+  }
+  return entityId;
+}
+
+async function enrichRelatedDids(row: EventRow, meta: EventMeta, dids: Set<string>): Promise<string | undefined> {
+  return await resolveEntityIdFromTxLocalData(row, meta);
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/services/api/indexer_events_query.ts]: Agree the name was misleading. We’re intentionally not using history-table enrichment, so I removed the unused dids param and just kept entity-id resolution.

File: src/services/api/indexer_events_query.ts [hunk 4]
```diff
@@ -443,6 +627,8 @@ async function buildIndexerTxEvents(args: {
       "tm.type as message_type",
       "tm.sender",
       "tm.content",
+      "tx.data as tx_data",
+      knex.raw("(select count(*)::int from transaction_message tm2 where tm2.tx_id = tx.id) as tx_message_count"),
       "tx.height as block_height",
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/services/api/indexer_events_query.ts:806]: Good call. I replaced the per-row correlated subquery with a window count(*) over (partition by tx.id).

File: src/services/api/indexer_events_query.ts [hunk 5]
```diff
@@ -473,12 +659,23 @@ export async function persistIndexerEventsForBlock(blockHeight: number): Promise
...
+        payload: knex.raw(
+          `
+          CASE
+            WHEN (indexer_events.payload->>'entity_id') IS NULL AND EXCLUDED.entity_id IS NOT NULL
+              THEN jsonb_set(COALESCE(indexer_events.payload, '{}'::jsonb), '{entity_id}', to_jsonb(EXCLUDED.entity_id::text), true)
+            ELSE COALESCE(indexer_events.payload, '{}'::jsonb)
+          END
+          `
+        ),
+      })
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/services/api/indexer_events_query.ts]: Makes sense. I added a WHERE so we only update/return rows that actually need backfilling, avoiding duplicate re-broadcasts.

File: src/migrations/20250905_create_trust_registry_tables.ts [hunk 6]
```diff
@@ -12,6 +12,8 @@ export async function up(knex: Knex): Promise<void> {
     table.string("language", 2).notNullable();
     table.integer("active_version").nullable();
     table.bigInteger("height").notNullable().unique();
+    table.unique(["did", "height"], "trust_registry_did_height_unique");
+    table.index(["height"], "trust_registry_height_idx");
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply src/migrations/20250905_create_trust_registry_tables.ts]: Yep height is already unique, so the extra constraint/index were redundant. I removed them.

File: test/unit/services/api/indexer_events_query.spec.ts [hunk 7]
```diff
@@ -130,8 +136,458 @@ describe("indexer_events_query", () => {
...
+              { key: "self_permission_id", value: "903" },
+            ],
+          },
+        ],
+      },
+    });
+    await persistIndexerEventsForBlock(baseHeight + 86);
+    const events = await listIndexerEvents({ did, blockHeight: baseHeight + 86, limit: 10 });
+    expect(events[0].payload.entity_id).toBeUndefined();
+    expect((global as any).logger.warn).toHaveBeenCalled();
```
[the remark this answers is not in the document]
Comment (mabubakr1113) [reply test/unit/services/api/indexer_events_query.spec.ts]: Wrapped it in try/finally to restore the previous (global as any).logger.

File: src/services/api/indexer_events_query.ts [hunk 8]
```diff
@@ -196,21 +202,363 @@ function readNumber(content: unknown, keys: string[]): number | null {
...
+    }
+  };
+
+  const buildAttrMap = (ev: TxEvent): Record<string, string> => {
+    const attrs = Array.isArray(ev?.attributes) ? (ev.attributes as TxEventAttribute[]) : [];
+    const attrMap: Record<string, string> = {};
+    for (const a of attrs) {
+      const k = toLowerSnake(decodeMaybeBase64((a as any)?.key));
+      if (!k) continue;
+      attrMap[k] = decodeMaybeBase64((a as any)?.value);
```
Comment (lotharking) [inline src/services/api/indexer_events_query.ts:469]: If possible, I'd prefer to avoid `any` typing

File: src/services/api/indexer_events_query.ts [hunk 9]
```diff
@@ -196,21 +202,363 @@ function readNumber(content: unknown, keys: string[]): number | null {
...
+
+  const normalizedAction = toLowerSnake(action);
+
+  const ruleByAction: Record<
+    string,
+    { eventTypes: string[]; idKeys: string[]; preferEventTypes?: string[]; allowEventTypeFallback?: boolean }
+  > = {
+    create_new_trust_registry: {
+      eventTypes: ["create_trust_registry", "create_new_trust_registry"],
+      idKeys: ["trust_registry_id", "tr_id", "id"],
```
Comment (lotharking) [inline src/services/api/indexer_events_query.ts:283]: Why are there so many hardcoded literals in this implementation? Please consider extracting them into constants, enums, or mapping objects to improve maintainability and reduce the risk of inconsistencies or typos.

File: src/services/api/indexer_events_query.ts [hunk 10]
```diff
@@ -196,21 +202,363 @@ function readNumber(content: unknown, keys: string[]): number | null {
...
+    .replace(/__+/g, "_")
+    .toLowerCase()
+    .trim();
+}
+
+function looksLikeBase64(s: string): boolean {
+  return /^[A-Za-z0-9+/=]+$/.test(s) && s.length % 4 === 0;
+}
+
+let warnedBase64DecodeFailure = false;
```
Comment (genaris) [inline src/services/api/indexer_events_query.ts:226]: Why using a global variable for this? Is it because you want to log the warning only once in an entire indexer session?

Looks pretety weird to me.

File: src/services/api/indexer_events_query.ts (hunk 8 above)
Comment (genaris) [inline src/services/api/indexer_events_query.ts:469]: This is so complex! Why is this processing that hard? "decode _maybe_  base 64?" What do you mean by maybe? And if the result of toLowerSnake is defined you call it again?

There must be an explanation for this but certainly it seems that it is adding a lot of noise to an already complex code like the one we have in the indexer.

**Review by genaris — CHANGES_REQUESTED** [review]
I didn't test it and surely it works fine, but honestly I don't see why there are pieces of code so intrincated. I'm worried about the maintenability of overall indexer code, and this makes it harder to understand.

If you have a clear explanation about why this direction is followed, let me know.

**Comment by genaris** [comment]
Superseded by #248
